### PR TITLE
Adding IWDS07 protocol (door/window sensor)

### PR DESCRIPTION
**Pull request by 1technophile**
Linked to #438 
And following this discussion
https://forum.pilight.org/showthread.php?tid=2952

### CODE REVIEW

**Comment by CurlyMoo** [comment]
So, you started a new PR and it's already a mess 😕 

I fixed the last PR, check this link how you can fix the current PR:
http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

**Comment by 1technophile** [comment]
Was automatically closed by github as there was no diff when I did the sync between branches.

Hope this PR is more suitable.

**Comment by CurlyMoo** [comment]
This is how it should be done.

File: libs/pilight/protocols/433.92/iwds07.c [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/*
+	Copyright (C) 2017 CurlyMo & wo-rasp
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Please include yourself in the author list and change the license to MPLv2

File: libs/pilight/protocols/433.92/iwds07.c [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	state = binToDec(binary, 21, 21);
+	createMessage(unit, battery, state);
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void iwds07Init(void) {
+	protocol_register(&iwds07);
+	protocol_set_id(iwds07, "GS-iwds07");
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Capitals or no capitals, that's the question.

File: libs/pilight/protocols/433.92/iwds07.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/*
+	Copyright (C) 2017 CurlyMo & wo_rasp
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.h]: Likewise

File: libs/pilight/protocols/433.92/iwds07.c [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	iwds07->hwtype = RF433;
+	iwds07->minrawlen = RAW_LENGTH;
+	iwds07->maxrawlen = RAW_LENGTH;
+	iwds07->maxgaplen = FOOTER-200;
+	iwds07->mingaplen = FOOTER+200;
+
+	options_add(&iwds07->options, "u", "unit", OPTION_HAS_VALUE, DEVICES_ID, JSON_NUMBER, NULL, NULL);
+	options_add(&iwds07->options, "b", "battery", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_NUMBER, NULL, "^[01]$");
+	options_add(&iwds07->options, "t", "opened", OPTION_NO_VALUE, DEVICES_STATE, JSON_STRING, NULL, NULL);
+	options_add(&iwds07->options, "f", "closed", OPTION_NO_VALUE, DEVICES_STATE, JSON_STRING, NULL, NULL);
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Where is the temper option?

File: libs/pilight/protocols/433.92/iwds07.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+static void createMessage(int unit, int battery, int state) {
+	iwds07->message=json_mkobject();
+	json_append_member(iwds07->message, "unit", json_mknumber(unit, 0));
+	json_append_member(iwds07->message, "battery", json_mknumber(battery, 0));
+	if(state == 1) {
+		json_append_member(iwds07->message, "state", json_mkstring("closed"));
+	} else {
+		json_append_member(iwds07->message, "state", json_mkstring("opened"));
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: I miss the temper option here as well.

File: libs/pilight/protocols/433.92/iwds07.c (hunk 4 above)
Comment (1technophile) [reply libs/pilight/protocols/433.92/iwds07.c]: I don't have a temper option on mine

File: libs/pilight/protocols/433.92/iwds07.c (hunk 5 above)
Comment (1technophile) [reply libs/pilight/protocols/433.92/iwds07.c]: likewise

Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/iwds07.c]: But why document a temper function when the protocol doesn't support it?

Comment (1technophile) [reply libs/pilight/protocols/433.92/iwds07.c]: removed from documentation

File: docs/source/protocols/433.92/contact/gs-iwds07.rst [hunk 6]
```diff
@@ -37,11 +37,12 @@ IWDS07 Contact
    {
      "devices": {
        "contact": {
-         "protocol": [ "IWDS07" ],
+         "protocol": [ "iwds07" ],
          "id": [{
            "unitcode": 0
          }],
          "state": "closed"
+         "battery":1
```
Comment (CurlyMoo) [inline docs/source/protocols/433.92/contact/gs-iwds07.rst]: Add a space between the colon and the one

File: docs/source/protocols/433.92/contact/gs-iwds07.rst [hunk 7]
```diff
@@ -0,0 +1,56 @@
+.. |yes| image:: ../../../images/yes.png
+.. |no| image:: ../../../images/no.png
+
+.. role:: underline
+   :class: underline
+
+IWDS07 Contact
+==================
```
Comment (CurlyMoo) [inline docs/source/protocols/433.92/contact/gs-iwds07.rst]: Can you still check this line.

File: docs/source/protocols/433.92/contact/gs-iwds07.rst [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        }
+     }
+   }
+
++------------------+--------------------------+
+| **Option**       | **Value**                |
++------------------+--------------------------+
+| id               | 0 - 524288               |
++------------------+--------------------------+
+| state            | opened / closed          |
```
Comment (CurlyMoo) [inline docs/source/protocols/433.92/contact/gs-iwds07.rst]: The battery option is missing in the table.

File: docs/source/protocols/433.92/contact/gs-iwds07.rst (hunk 7 above)
Comment (CurlyMoo) [reply docs/source/protocols/433.92/contact/gs-iwds07.rst]: I wrote before that the equal signs should be the same length of the title.

File: CMakeConfig.txt [hunk 9]
```diff
@@ -42,6 +42,7 @@ set(PROTOCOL_HEITECH ON CACHE BOOL "support for the Heitech switch protocol")
 set(PROTOCOL_IMPULS ON CACHE BOOL "support for the Impuls switch protocol")
 set(PROTOCOL_INTERTECHNO_SWITCH ON CACHE BOOL "support for the new Intertechno switch protocol")
 set(PROTOCOL_INTERTECHNO_OLD ON CACHE BOOL "support for the old Intertecho switch protocol")
+set(PROTOCOL_IWDS07 ON CACHE BOOL "support for the IWDS07 contact protocol")
```
Comment (CurlyMoo) [inline CMakeConfig.txt]: Can you add the brand `Golden Security` here.

File: docs/source/protocols/433.92/contact/gs-iwds07.rst [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
++------------------+-------------+
+| Config           | |yes|       |
++------------------+-------------+
+
+.. rubric:: Supported Brands
+
++------------------+--------------+
+| **Brand**        | **Protocol** |
++------------------+--------------+
+| GS               | iwds07       |
```
Comment (CurlyMoo) [inline docs/source/protocols/433.92/contact/gs-iwds07.rst]: And here.

File: docs/source/protocols/433.92/contact/gs-iwds07.rst [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+.. rubric:: Sender Arguments
+
+*None*
+
+.. rubric:: Config
+
+.. code-block:: json
+   :linenos:
+
+   {
```
Comment (CurlyMoo) [inline docs/source/protocols/433.92/contact/gs-iwds07.rst]: The example json is invalid

File: libs/pilight/protocols/433.92/iwds07.c [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+	iwds07->mingaplen = FOOTER+200;
+
+	options_add(&iwds07->options, "u", "unit", OPTION_HAS_VALUE, DEVICES_ID, JSON_NUMBER, NULL, NULL);
+	options_add(&iwds07->options, "b", "battery", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_NUMBER, NULL, "^[01]$");
+	options_add(&iwds07->options, "t", "opened", OPTION_NO_VALUE, DEVICES_STATE, JSON_STRING, NULL, NULL);
+	options_add(&iwds07->options, "f", "closed", OPTION_NO_VALUE, DEVICES_STATE, JSON_STRING, NULL, NULL);
+
+	iwds07->parseCode=&parseCode;
+	iwds07->validate=&validate;
+}
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Add a newline here

**Comment by CurlyMoo** [comment]
Now, rebase the whole thing again to one commit and you've learned how to properly do PR's :)

**Comment by 1technophile** [comment]
:-) yes, it was long but instructive!

**Comment by 1technophile** [comment]
Thanks

**Comment by CurlyMoo** [comment]
Likewise

**Pull request merged**